crates/iota-grpc-api/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{pin::Pin, sync::Arc};
+
+use serde::{Deserialize, Serialize};
+use tonic::{Request, Response, Status};
+
+/// Configuration for the gRPC API service
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct Config {
+    /// The address to bind the gRPC server to
+    #[serde(default = "default_grpc_api_address")]
+    pub address: std::net::SocketAddr,
+
+    /// Buffer size for broadcast channels used for checkpoint streaming
+    #[serde(default = "default_checkpoint_broadcast_buffer_size")]
+    pub checkpoint_broadcast_buffer_size: usize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            address: default_grpc_api_address(),
+            checkpoint_broadcast_buffer_size: default_checkpoint_broadcast_buffer_size(),
+        }
+    }
+}
+
+fn default_grpc_api_address() -> std::net::SocketAddr {
+    use std::net::{IpAddr, Ipv4Addr};
+    std::net::SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 50051)
+}
+
+fn default_checkpoint_broadcast_buffer_size() -> usize {
+    100
+}
+
+pub mod node {
+    tonic::include_proto!("iota.grpc");
+}
+
+use iota_types::storage::RestStateReader;
+use node::{BcsData, node_service_server::NodeService};
+pub mod client;
+use iota_grpc_types::{
+    CertifiedCheckpointSummary as GrpcCertifiedCheckpointSummary,
+    CheckpointData as GrpcCheckpointData,
+};
+use iota_types::{
+    full_checkpoint_content::CheckpointData, messages_checkpoint::CertifiedCheckpointSummary,
+};
+use tracing::{debug, info};
+
+type Receiver<T> = tokio::sync::broadcast::Receiver<T>;
+
+impl BcsData {
+    fn serialize_from<T>(data: &T) -> Result<Self, bcs::Error>
+    where
+        T: Serialize,
+    {
+        let serialized = bcs::to_bytes(data)?;
+        Ok(BcsData { data: serialized })
+    }
+
+    fn deserialize_into<T>(&self) -> Result<T, bcs::Error>
+    where
+        T: for<'de> Deserialize<'de>,
+    {
+        bcs::from_bytes(&self.data)
+    }
+}
+
+pub struct NodeGrpcService {
+    pub state_reader: Arc<dyn RestStateReader>,
```

Comment:
Why don't you hold a Reader here directly instead of always creating one from the value of this field?